### DISCUSSION TITLE
if_id_in/out not incrementing

### DISCUSSION THREAD

**Discussion by cyanideangel2000** [strongswan/strongswan#1398] [Q&A] [answered]
hello Team,

I am running into a weird issue.

I have a set of tunnels configured via legacy ipsec.conf. These tunnels use mark_in/mark_out and netfilter iptable rules for traffic redirection.
Everything works fine.

Now, I am migrating these to xfrm if_id_in/out based semantics.
I have now configured these tunnels via swanctl.conf with if_id_in/out set as %unique.
Everything works fine.

The problem.
I have a switch which when enabled migrates the tunnel from ipsec.conf mark based semantics to swanctl xfrm if_id based semantics.
When i start the service, its still ipsec starter based therefore tunnels gets loaded via ipsec.conf. I can see in the logs about stroke.
Tunnel comes up and traffic flows.
Now I terminate these tunnels and enable the switch to migrate to xfrm if_id semantics.
I trigger a swanctl --load-all as an action of enabling the switch.
I can see vici connections getting loaded.

Now when I bring up the tunnels, xfrm if_id remains 0 for in and out and for all tunnels.
and the problem is specific to tunnels which exist both in ipsec.conf and swanctl.conf. If a tunnel only exist in swanctl.conf, it just works fine.

strongswan 5.9.7.
Any thoughts?

**Comment by tobiasbrunner** [comment, marked as answer]
I don't get what you are doing. What's the point of configuring the same connection in ipsec.conf and again in swanctl.conf and then do some "switching" at runtime? But anyway, you obviously shouldn't name them the same because names are global for some contexts (e.g. initiating them), so even when using swanctl/vici to initiate a connection this might initiate the one configured via ipsec.conf (unless you unload it first), which obviously won't allocate any if_ids.

**Reply by cyanideangel2000** [reply]
Hello Tobias,

Thank you for the insight.
I understand  that having the same tunnel defined into via both ipsec.conf and swanctl isn't the best thing, however I do have a transient need for the same.
I didn't clarify the box where I trying to debug this is the Ipsec responder.
Tunnels are initiated by a different entity and nothing changes at its end irrespective.
I probed a bit into strongswan 5.9.7 code and didn't really find anything that should explain this.
xfrm if_id allocation is simply atomic increment and doesn't really cares about the any other thing.

**Reply by tobiasbrunner** [reply]
> however I do have a transient need for the same.

What might that be?

Anyway, if you have two equivalent connections loaded as responder (the name doesn't matter in that case), except for the if_id config, it could very well be that the one from ipsec.conf without if_id is chosen. While there is a "best" match, if two connections are equally valid based on IPs, IKE proposals and identities, it  will depend on the order in which plugins (stroke is loaded before vici in the default config) and connections are loaded.

**Reply by cyanideangel2000** [reply]
Thank you Tobias, i think you answered my question. I seem to understand why am i seeing what I am seeing.
Yes all tunnels are identical for me, and stroke gets loaded first.
